### PR TITLE
lxd native mount

### PR DESCRIPTION
**Pull request by georgeliao**
[JIRA story link](https://warthogs.atlassian.net/browse/MULTI-188)
[spek link](https://docs.google.com/document/d/1LAnT92_lhfsIQD_JIbb4JfsqVakRXVpJnFYT1gpNMCQ/edit#)
The current workflow and behaviors are as follows: 
`./multipass info peaceable-tadpole
`
```
Name:           peaceable-tadpole
State:          Running
IPv4:           10.156.2.154
Release:        Ubuntu 22.04.2 LTS
Image hash:     d900c08af2be (Ubuntu 22.04 LTS)
CPU(s):         1
Load:           0.00 0.00 0.00
Disk usage:     1.6GiB out of 9.5GiB
Memory usage:   190.9MiB out of 946.1MiB
Mounts:         --
```

`./multipass mount -t native /home/georgel/mount peaceable-tadpole:/home/ubuntu/native
`
`mount failed: Please stop the instance peaceable-tadpole before mount it natively.
`

`./multipass stop peaceable-tadpole
`

`./multipass mount -t native /home/georgel/mount peaceable-tadpole:/home/ubuntu/native
`

`./multipass start peaceable-tadpole
`
`./multipass exec peaceable-tadpole bash 
`
`cd native`
Then you can use the shared directory
`exit`

`/multipass unmount peaceable-tadpole:/home/ubuntu/native
`
Here is where it gets interesting, we should not allow hot unplugging, so there are two options for the behaviors. 

1. Throw an error to ask the user to manually stop the instance
2. The unmount does this automatically, that is stopping the instance first if it is still running, then do the unmount.

The current implementation is based on 2nd behavior.

`./multipass info peaceable-tadpole
`
```
Name:           peaceable-tadpole
State:          Stopped
IPv4:           --
Release:        --
Image hash:     d900c08af2be (Ubuntu 22.04 LTS)
CPU(s):         --
Load:           --
Disk usage:     --
Memory usage:   --
Mounts:         --
```

### CODE REVIEW

**Comment by ricab** [comment]
Hey @georgeliao, thank you for this! Concerning hot unmounting, I don't think we should take the initiative to stop the instance. Users could have things running and unsaved work. So, if we can't do better, I think we should go with option 1 and error out. 

Something else we could try (let's call it option 3): manually unmount inside the instance and later update things with LXD, when the instance next stops or starts. That could be a little tricky though, I am not sure it's worth the trouble right now...

I have only had a very superficial look at the spec yet, but I didn't see this topic discussed? I think this should be there, and it may tie in with the discussion for PUT/PATCH (e.g. you could use PUT in a state transition to basically "push" Multipass's internal view of mounts to LXD).

**Comment by georgeliao** [comment]
@ricab 
Thanks for thinking of option 3. It can be an alternative.
In my opinion, option 1 is the best because it is symmetric to hot-mount. However, there is an obstacle to the implementation though. Because the destructor of LXDMountHandler handles un-mount and throwing an error in the destructor is not allowed. There are other alternatives though, but that will make the code messier. 

About the spec, yes, this topic is not covered because I did notice this when I was writing it. I can add it in as well.

**Comment by ricab** [comment]
> In my opinion, option 1 is the best because it is symmetric to hot-mount. 

I see your point, but then it is asymmetric with respect to hot-unmount in QEMU. Usage then really starts to vary a lot depending on the backend.

> [...] the destructor of LXDMountHandler handles un-mount [...]

Yeah, I see what you mean. IMO, the strongest reason to go with option 1 instead of 3 would be simplicity. But this means it is not so simple anymore.

> About the spec, yes, this topic is not covered because I did notice this when I was writing it. I can add it in as well.

Sure, that's fine :slightly_smiling_face: You can add something brief once the way forward is clear. Let's discuss in the sync later.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        }
+        mpl::log(mpl::Level::warning, std::string(category),
+                 fmt::format("Failed to gracefully stop mount \"{}\" in instance '{}': {}", target, vm->vm_name,
+                             general_exception.what()));
+    }
+}
+
+LXDMountHandler::~LXDMountHandler()
+{
+    stop(true);
```
Comment (ricab) [inline src/platform/backends/lxd/lxd_mount_handler.cpp:81]: This would still need wrapping in a try-catch.

File: src/platform/backends/lxd/lxd_virtual_machine.cpp [hunk 2]
```diff
@@ -459,3 +470,15 @@ void mp::LXDVirtualMachine::resize_disk(const MemorySize& new_size)
...
     lxd_request(manager, "PATCH", url(), patch_json);
 }
+
+std::unique_ptr<multipass::MountHandler>
+mp::LXDVirtualMachine::make_native_mount_handler(const SSHKeyProvider* ssh_key_provider, const std::string& target,
+                                                 const VMMount& mount)
+{
+    if (!is_default_Gid_Uid(mount))
+    {
+        throw std::runtime_error("LXD native mount does not accept user provided gid or uid.");
```
Comment (ricab) [inline src/platform/backends/lxd/lxd_virtual_machine.cpp:480]:
```suggestion
        throw std::runtime_error("LXD native mount does not accept custom gid or uid.");
```

This class does not know where IDs come from. We know they come from the user but, as far as this class is concerned, they could also come from a custom client or whatever.

File: include/multipass/mount_handler.h [hunk 3]
```diff
@@ -63,6 +63,11 @@ class MountHandler : private DisabledCopyMove
         return active;
     }
 
+    virtual bool is_sticky()
+    {
```
[the remark this answers is not in the document]
Comment (ricab) [reply include/multipass/mount_handler.h:67]: One alternative that occurred to me was `is_managed`, as in "managed by the backend". But I thought "sticky" was more descriptive of the behavior: the mount sticks to the VM, irrespective of state transitions or the Multipass daemon.

File: tests/lxd/test_lxd_mount_handler.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    mp::LXDMountHandler lxd_mount_handler(&mock_network_access_manager, &lxd_vm, &key_provider, target_path, vm_mount);
+
+    EXPECT_CALL(lxd_vm, current_state).WillOnce(Return(multipass::VirtualMachine::State::running));
+    const mp::ServerVariant dummy_server;
+    MP_EXPECT_THROW_THAT(
+        lxd_mount_handler.activate(dummy_server), std::runtime_error,
+        mpt::match_what(AllOf(HasSubstr("Please stop the instance"), HasSubstr("before mount it natively."))));
+}
+
+TEST_F(LXDMountHandlerTestFixture, stopNoThrowsIfVMIsStopped)
```
Comment (ricab) [inline tests/lxd/test_lxd_mount_handler.cpp:136]:
```suggestion
TEST_F(LXDMountHandlerTestFixture, stopDoesNotThrowIfVMIsStopped)
```

**Review by ricab — CHANGES_REQUESTED** [review]
Hey @georgeliao, nicely done! I have a bunch of smaller requests, mostly on naming and language, as well as some missing coverage that should be really easy to get. Also a few code improvements, but your code and approach are fundamentally sound.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+#include "lxd_mount_handler.h"
+#include "lxd_request.h"
+
+namespace
+{
+constexpr std::string_view category = "lxd-mount-handler";
+constexpr int length_of_unique_id_without_prefix = 25;
+constexpr int timeout_millisecond = 300000;
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:25]: The time-out value is something I did not think about, I simply copied from another lxd_wait example somewhere else. But indeed, depending on the action he is waiting for, the value can be different, I can change that of course. About the variable name, my intention is to mean 'timeout' (type of value), 'millisecond' (unit), like "timeout in millisecond" or "distance in meter". I can change it to "timeout_in_millisecond" or "timeout_milliseconds" as you suggested.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+{
+LXDMountHandler::LXDMountHandler(mp::NetworkAccessManager* network_manager, LXDVirtualMachine* lxd_virtual_machine,
+                                 const SSHKeyProvider* ssh_key_provider, const std::string& target_path,
+                                 const VMMount& mount)
+    : MountHandler{lxd_virtual_machine, ssh_key_provider, target_path, mount.source_path},
+      network_manager_{network_manager},
+      lxd_instance_endpoint_{
+          QString("%1/instances/%2").arg(lxd_socket_url.toString()).arg(lxd_virtual_machine->vm_name.c_str())},
+      // 27 (25 + 2(d_)) letters is the maximal deivce name length that lxd can accepte
+      device_name_{
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:38]: I know we have been calling it device id during our conversations, however, I think it is device name by lxd's definition 
image

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:25]: I think it is a good choice to include the unit in the variable name. But I am pretty sure that the correct English is "distance in meter**s**" and "timeout in millisecond**s**" (the distance is measured in (potentially) multiple meters, the timeout consists of (potentially) multiple milliseconds). But perhaps a native English speaker can correct me.

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:38]: OK, fair enough.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+          mp::utils::make_uuid(target_path).left(length_of_unique_id_without_prefix).prepend("d_").toStdString()}
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:39]: I would like to follow the nomenclature 'devce_<hash>', since 'device' alone took too many digits, so I have shortened it to 'd_', the intention is to make it indicative that it is a device name.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+namespace multipass
+{
+LXDMountHandler::LXDMountHandler(mp::NetworkAccessManager* network_manager, LXDVirtualMachine* lxd_virtual_machine,
+                                 const SSHKeyProvider* ssh_key_provider, const std::string& target_path,
+                                 const VMMount& mount)
+    : MountHandler{lxd_virtual_machine, ssh_key_provider, target_path, mount.source_path},
+      network_manager_{network_manager},
+      lxd_instance_endpoint_{
+          QString("%1/instances/%2").arg(lxd_socket_url.toString()).arg(lxd_virtual_machine->vm_name.c_str())},
+      // 27 (25 + 2(d_)) letters is the maximal deivce name length that lxd can accepte
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:37]: Nice comment, I did look into the difference between 'maximal' and 'maximum', indeed, 'maximum' is more appropriate here.

File: src/platform/backends/lxd/lxd_virtual_machine.cpp [hunk 9]
```diff
@@ -153,6 +154,16 @@ QJsonObject generate_devices_config(const multipass::VirtualMachineDescription&
     return devices;
 }
 
+bool is_default_Gid_Uid(const multipass::VMMount& mount)
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_virtual_machine.cpp:157]: Good catch, indeed the name "is_default_Gid_Uid" makes more sense if the input parameter is gid and uid map. uses_default_id_mappings can be more appropriate when the mount is the parameter.

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:39]: I see... 'd_' is a little cryptic though, I am not sure it helps the reader at all. However, LXD might need the first character to be alphabetical. That problem was already solved for the QEMU case, I would think an analogous approach would be best here.

File: tests/lxd/test_lxd_mount_handler.cpp [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+#include <multipass/utils.h>
+#include <multipass/virtual_machine_description.h>
+#include <multipass/vm_mount.h>
+
+namespace mp = multipass;
+namespace mpt = multipass::test;
+
+namespace
+{
+class MockLXDVirtualMachine : public mp::LXDVirtualMachine
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply tests/lxd/test_lxd_mount_handler.cpp:37]: Yep, will do. However, I would say one small advantage of the current inheritance is that the MOCK_METHOD list in the MockLXDVirtualMachine shows exactly what he needs, it is just `current_state()` function in this case. In the uniform MockVirtualMachineT case, it mocked a lot of things for different client classes even though some are not needed by some clients.

File: include/multipass/mount_handler.h [hunk 11]
```diff
@@ -35,26 +35,35 @@ using ServerVariant = std::variant<grpc::ServerReaderWriterInterface<StartReply,
...
+
 class MountHandler : private DisabledCopyMove
 {
 public:
     using UPtr = std::unique_ptr<MountHandler>;
 
     virtual ~MountHandler() = default;
 
-    void start(ServerVariant server, std::chrono::milliseconds timeout = std::chrono::minutes(5))
+    void activate(ServerVariant server, std::chrono::milliseconds timeout = std::chrono::minutes(5))
```
Comment (georgeliao) [inline include/multipass/mount_handler.h:54]: TODO, windows side SMPMountHandler name change as well.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+void LXDMountHandler::lxd_device_remove()
+{
+    const QJsonObject instance_info = lxd_request(network_manager_, "GET", lxd_instance_endpoint_);
+    QJsonObject instance_info_metadata = instance_info["metadata"].toObject();
+    QJsonObject device_list = instance_info_metadata["devices"].toObject();
+
+    device_list.remove(device_name_.c_str());
+    instance_info_metadata["devices"] = device_list;
+
+    const QJsonObject json_reply = lxd_request(network_manager_, "PUT", lxd_instance_endpoint_, instance_info_metadata);
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:83]: I will look into this later, it might take a while, so it can be a separated PR.

File: src/daemon/daemon.cpp [hunk 13]
```diff
@@ -1672,11 +1672,11 @@ try // clang-format on
 
         VMMount vm_mount{request->source_path(), gid_mappings, uid_mappings, mount_type};
         vm_mounts[target_path] = make_mount(vm.get(), target_path, vm_mount);
-        if (vm->current_state() == mp::VirtualMachine::State::running)
+        if (vm->current_state() == mp::VirtualMachine::State::running || !vm_mounts[target_path]->is_sticky())
         {
             try
             {
-                vm_mounts[target_path]->start(server);
+                vm_mounts[target_path]->activate(server);
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/daemon/daemon.cpp:1885]: A rebase seemed to resolve it, now the patch coverage is 94.59%

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:41]: Right. With the comment there explaining that the first char needs to be alphanumerical, one quickly understands why the `m` is there, and that it must stand for "mount". 

I would have liked to see a consistent approach in both cases, so that I don't have to re-understand one if I already understood the other. Consistency is definitely not the most important value, but it is not valueless either.

If the two cases were equivalent, you could also change the QEMU case. However, in this respect I find that the QEMU implementation is slightly better because:

1. it explains why something is prepended at all; 
2. it is obvious what the prefix stands for;
3. this looses an extra UUID digit to have the underscore (which is important here since `d` is also a hex digit).

I am not going to block on this detail though.

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:83]: Yeah, I am OK with a separate PR for this one as discussed elsewhere.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    mpl::log(mpl::Level::info, std::string(category),
+             fmt::format("initializing native mount {} => {} in '{}'", source, target, vm->vm_name));
+    lxd_device_add();
+}
+
+void LXDMountHandler::deactivate_impl(bool /*force*/)
+{
+    // TODO: remove the throwing and adjust the unit tests once lxd fix the hot-unmount bug
```
Comment (ricab) [inline src/platform/backends/lxd/lxd_mount_handler.cpp:60]: Yep, that's what I meant :slightly_smiling_face:

File: tests/lxd/test_lxd_mount_handler.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ricab) [reply tests/lxd/test_lxd_mount_handler.cpp]: I see that rather as a problem though. If the tests assume that is the only thing that needs to be mocked, they are relying on the tested code to be correct in that way, i.e., that it doesn't use anything else. Otherwise, tests run production code. 

IOW, tests should work safely against wrong code too, including code calling other methods. If the methods are mocked, that is caught and dealt with according on the personality of the mock (strict,naggy,nice) and production code is not run.

File: src/platform/backends/lxd/lxd_mount_handler.h [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+private:
+    void lxd_device_add();
+    void lxd_device_remove();
+
+    // data memeber
+    mp::NetworkAccessManager* network_manager_{nullptr};
+    const QUrl lxd_instance_endpoint_{};
+    const std::string device_name_{};
```
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/lxd/lxd_mount_handler.h:48]: OK, I am supportive of the idea, provided we do it right. Let's discuss this with the others then.

File: tests/mock_virtual_machine.h [hunk 16]
```diff
@@ -33,7 +33,8 @@ namespace test
 template <typename T = VirtualMachine, typename = std::enable_if_t<std::is_base_of_v<VirtualMachine, T>>>
 struct MockVirtualMachineT : public T
 {
-    MockVirtualMachineT(const std::string vm_name) : T{vm_name}
+    template <typename... Args>
+    MockVirtualMachineT(Args&&... args) : T{args...}
```
Comment (ricab) [inline tests/mock_virtual_machine.h:37]: Nice, this was missing indeed. But may as well use perfect forwarding:

```suggestion
    template <typename... Args>
    MockVirtualMachineT(Args&&... args) : T{std::forward<Args>(args)...}
```

**Review by ricab — COMMENTED** [review]
Hey @georgeliao, a few remaining details.

File: include/multipass/mount_handler.h [hunk 17]
```diff
@@ -35,26 +35,35 @@ using ServerVariant = std::variant<grpc::ServerReaderWriterInterface<StartReply,
                                    grpc::ServerReaderWriterInterface<MountReply, MountRequest>*,
                                    grpc::ServerReaderWriterInterface<RestartReply, RestartRequest>*>;
 
+class MountHandlerActivateException : public std::runtime_error
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply include/multipass/mount_handler.h:38]: When I was naming it, I did not thought about other possible exceptions in the `activate()`, but indeed, that name is too general, `NativeMountNeedsStoppedVMException` is better.

File: tests/mock_virtual_machine.h (hunk 16 above)
Comment (georgeliao) [reply tests/mock_virtual_machine.h:37]: Yep, you are right, can not believe I have overlooked that. Universal reference `Args&& `does retain value type in general. However, when rvalue `args` get into the function (`MockVirtualMachineT` constructor in this case), it gets a name, a variable with a name becomes lvalue. So the `args` passed to sub-function (`T` constructor in this case) without `std::forward` will be lvalue instead of rvalue, which might cause either wrong constructor match or can not find the wanted constructor.

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:41]: LXD does not really require the first letter to be alphabetic, adding d_ is just for the sake of indicating it is a device name. What the prefix stands for (device) is different from QEMU case, so I think it is ok to be a different letter. Removing dash can be applied here as well because it adds more digits in a way, the reason I did not do it is because it leads to more string manipulation which means more time cost, however, it might be small in this case.

To summarize, we can leave the first letter and length to be different because they are different in these two cases.

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:87]: Will do.

**Review by ricab — COMMENTED** [review]
Hey @georgeliao, the code is looking good, modulo the private side and the field naming question. However, I found some weird behavior when testing mounts with relative paths. This appears to be mostly a bug in LXD, but I am not sure. We may be able to workaround on our side. Look at this session:

```
ricab@ricab-laptop:~ $ pwd
/home/ricab
ricab@ricab-laptop:~ $ mkdir tmp
ricab@ricab-laptop:~ $ echo asdf > ./tmp/asdf
ricab@ricab-laptop:~ $ mp get local.driver
lxd
ricab@ricab-laptop:~ $ mp launch -n t
Launched: t
ricab@ricab-laptop:~ $ mp stop t
ricab@ricab-laptop:~ $ mp mount -t native ~/tmp t:./tmp
ricab@ricab-laptop:~ $ ls tmp
asdf
ricab@ricab-laptop:~ $ mp info t
Name:           t
State:          Stopped
IPv4:           --
Release:        --
Image hash:     2750ece0097a (Ubuntu 22.04 LTS)
CPU(s):         --
Load:           --
Disk usage:     --
Memory usage:   --
Mounts:         /home/ricab/tmp => tmp
                    UID map: 1000:default
                    GID map: 1000:default
ricab@ricab-laptop:~ $ # so far so good
ricab@ricab-laptop:~ $ mp exec -n t -- ls -la tmp
ls: cannot access 'tmp': No such file or directory
ricab@ricab-laptop:~ $ lxc --project=multipass config device show t | grep -A3 d_
d_d98a3dba255a39c1b6875e16b:
  path: tmp
  source: /home/ricab/tmp
  type: disk
ricab@ricab-laptop:~ $ mp exec -n t -- grep d_d98a3dba255a39c1b6875e16b /proc/mounts
lxd_d_d98a3dba255a39c1b6875e16b /tmp virtiofs rw,relatime 0 0
ricab@ricab-laptop:~ $ # ugh... /tmp??
ricab@ricab-laptop:~ $
ricab@ricab-laptop:~ $ ls -l tmp
total 16
drwx------ 3 root root 4096 jun  6 12:59 snap-private-tmp
drwx------ 3 root root 4096 jun  6 12:59 systemd-private-54ca258146c446c7932f3739ddf36bec-systemd-logind.service-sADL6u
drwx------ 3 root root 4096 jun  6 12:59 systemd-private-54ca258146c446c7932f3739ddf36bec-systemd-resolved.service-BlwtJe
drwx------ 3 root root 4096 jun  6 12:59 systemd-private-54ca258146c446c7932f3739ddf36bec-systemd-timesyncd.service-7RBPGp
ricab@ricab-laptop:~ $ ls -la | grep tmp
drwxrwxrwt 11 root  root     4096 jun  6 12:59 tmp
ricab@ricab-laptop:~ $ # now my source directory was overwritten with lxd's stuff o_O
ricab@ricab-laptop:~ $ sudo tree -a tmp/
tmp/
├── .font-unix
├── .ICE-unix
├── snap-private-tmp
│   └── snap.lxd
│       └── tmp
├── systemd-private-54ca258146c446c7932f3739ddf36bec-systemd-logind.service-sADL6u
│   └── tmp
├── systemd-private-54ca258146c446c7932f3739ddf36bec-systemd-resolved.service-BlwtJe
│   └── tmp
├── systemd-private-54ca258146c446c7932f3739ddf36bec-systemd-timesyncd.service-7RBPGp
│   └── tmp
├── .Test-unix
├── .X11-unix
└── .XIM-unix

14 directories, 0 files
ricab@ricab-laptop:~ $
ricab@ricab-laptop:~ $ grep tmp /proc/mounts
udev /dev devtmpfs rw,nosuid,relatime,size=16170092k,nr_inodes=4042523,mode=755,inode64 0 0
tmpfs /run tmpfs rw,nosuid,nodev,noexec,relatime,size=3247176k,mode=755,inode64 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev,inode64 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k,inode64 0 0
tmpfs /run/snapd/ns tmpfs rw,nosuid,nodev,noexec,relatime,size=3247176k,mode=755,inode64 0 0
tmpfs /var/snap/lxd/common/ns tmpfs rw,relatime,size=1024k,mode=700,inode64 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=3247172k,nr_inodes=811793,mode=700,uid=1000,gid=1000,inode64 0 0
tmpfs /run/qemu tmpfs rw,nosuid,nodev,relatime,mode=755,inode64 0 0
ricab@ricab-laptop:~ $
ricab@ricab-laptop:~ $ # nothing matching /home/ricab/tmp, so it doesn't look like a bind mount either
ricab@ricab-laptop:~ $
ricab@ricab-laptop:~ $ mp umount t
umount failed: The following errors occurred:
failed to unmount "tmp" from 't': Please stop the instance t before unmount it natively.
ricab@ricab-laptop:~ $ mp stop t
ricab@ricab-laptop:~ $ mp umount t
ricab@ricab-laptop:~ $ mp info t | grep Mounts
Mounts:         --
ricab@ricab-laptop:~ $ ll tmp/
total 32
drwxrwxrwt  8 root  root  4096 jun  6 13:10 .
drwxr-x--- 40 ricab ricab 4096 jun  6 12:56 ..
drwxrwxrwt  2 root  root  4096 jun  6 12:59 .font-unix
drwxrwxrwt  2 root  root  4096 jun  6 12:59 .ICE-unix
drwx------  3 root  root  4096 jun  6 12:59 snap-private-tmp
drwxrwxrwt  2 root  root  4096 jun  6 12:59 .Test-unix
drwxrwxrwt  2 root  root  4096 jun  6 12:59 .X11-unix
drwxrwxrwt  2 root  root  4096 jun  6 12:59 .XIM-unix
ricab@ricab-laptop:~ $ # still the same
```

**Comment by georgeliao** [comment]
@ricab I will look into that.

File: src/platform/backends/lxd/lxd_mount_handler.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.h]: As we decided in our team meeting, it will be deferred, so I will change the field name back.

File: src/platform/backends/lxd/lxd_mount_handler.cpp [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    lxd_wait(network_manager, multipass::lxd_socket_url, json_reply, timeout_milliseconds);
+}
+
+void LXDMountHandler::lxd_device_add()
+{
+    const QJsonObject instance_info = lxd_request(network_manager, "GET", lxd_instance_endpoint);
+    QJsonObject instance_info_metadata = instance_info["metadata"].toObject();
+    QJsonObject device_list = instance_info_metadata["devices"].toObject();
+
+    const std::string abs_target_path = std::filesystem::path{target}.is_relative() ? "/home/ubuntu/" + target : target;
```
Comment (ricab) [inline src/platform/backends/lxd/lxd_mount_handler.cpp:97]: Ideally this should be tied to the user in `daemon_config.cpp`. I am going to overlook it though, because I am not sure how big a refactor that would require now and it's about time we get this in. If you find that's not that hard and want to deal with it in a separate PR (e.g. just pass it in within the `VMMount`?), that'd still be welcome :wink:

**Review by ricab — APPROVED** [review]
Right, that's it @georgeliao! Somehow it never occurred to me that it was mounting relative to the root.

Thanks for the PR, it is looking great! Especially considering your other follow-up PRs :slightly_smiling_face: 

bors r+

**Comment by ricab** [comment]
bors cancel

No point wasting cycles, waiting for the other side to be ready first.

File: src/platform/backends/lxd/lxd_mount_handler.cpp (hunk 18 above)
Comment (georgeliao) [reply src/platform/backends/lxd/lxd_mount_handler.cpp:97]: Ok, now I can put the pieces together, I sensed that `"ubuntu"` string should be a variable, but I did not know which one that was. 
I think `ssh_username` is accessible within `LXDMountHanlder` class, because `virutal_machine` class has the `ssh_username` method which calls `LXDVirtualMachine::ssh_username` method,

Comment (ricab) [reply src/platform/backends/lxd/lxd_mount_handler.cpp]: Ahh cool, even better. Well spotted! :+1:
